### PR TITLE
add validation for supported nstID

### PR DESCRIPTION
**Pull request by leonz789**
## Description

<p>Restrict current implementation to support NSTETH only.</p>
----

Closes #XXX

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of asset IDs for NSTETH with new constants and a mapping for maximum effective balances.
	- Introduced a new function to verify asset ID support for balance updates.

- **Bug Fixes**
	- Corrected the error variable name for unsupported asset IDs, ensuring consistency and clarity in error handling.
	- Updated logic for deposit and withdrawal handling to improve accuracy in validator management.

- **Refactor**
	- Improved structure and clarity in balance change processing methods, including the introduction of capped balance change handling.

### CODE REVIEW

File: x/oracle/keeper/native_token.go [hunk 1]
```diff
@@ -21,7 +21,26 @@ import (
 // msg(refund or slash on beaconChain): update staker's price, operator's price
 
 const (
-	NSTETHASSETID = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
+	// TODO: we currently support NSTETH only which has capped effective balance for one validator
+	// TODO: this is a bad practice, and for Lz, they have different version of endpoint with different chainID
+	// Do the validation before invoke oracle ralted functions instead of check these hard code ids here.
+	NSTETHAssetIDMainnet = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
```
Comment (cloud8little) [inline x/oracle/keeper/native_token.go]: mark 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65 as localnet. Mainnet should be 0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x7595

Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go [hunk 2]
```diff
@@ -392,3 +409,13 @@
...
+
+// IsLimitChangesNST returns that is input assetID corresponding to asset which balance change has a cap limit
+func IsLimitedChangeNST(assetID string) bool {
+	for _, aID := range limitedChangeNST {
+		if aID == assetID {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go [hunk 3]
```diff
@@ -21,7 +21,27 @@
...
+		NSTETHAssetIDMainnet,
+		NSTETHAssetIDHolesky,
+		NSTETHAssetIDSepolia,
+	}
+
+	maxEffectiveBalance = map[string]int{
+		NSTETHAssetIDMainnet: 32,
+		NSTETHAssetIDHolesky: 32,
+		NSTETHAssetIDSepolia: 32,
+	}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go [hunk 4]
```diff
@@ -244,6 +256,9 @@
 
 // UpdateNSTByBalanceChange updates balance info for staker under native-restaking asset of assetID when its balance changed by slash/refund on the source chain (beacon chain for eth)
 func (k Keeper) UpdateNSTByBalanceChange(ctx sdk.Context, assetID string, rawData []byte, roundID uint64) error {
+	if !IsLimitedChangeNST(assetID) {
+		return types.ErrNSTAssetNotSurpported
+	}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go [hunk 5]
```diff
@@ -133,6 +148,14 @@
 }
 
 func (k Keeper) UpdateNSTValidatorListForStaker(ctx sdk.Context, assetID, stakerAddr, validatorPubkey string, amount sdkmath.Int) error {
+	if !IsLimitedChangeNST(assetID) {
+		return types.ErrNSTAssetNotSurpported
+	}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: fixed

File: x/oracle/keeper/native_token.go [hunk 6]
```diff
@@ -146,7 +169,7 @@
 		stakerInfo = types.NewStakerInfo(stakerAddr, validatorPubkey)
 	} else {
 		k.cdc.MustUnmarshal(value, stakerInfo)
-		if amount.IsPositive() {
+		if amountInt64 > 0 {
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go:177]: one validator can only deposit once before it completed withdraw which remove its pubkey form this list. So there's no need to check duplication

File: x/oracle/keeper/native_token.go [hunk 7]
```diff
@@ -21,7 +21,29 @@ import (
...
 
 const (
-	NSTETHASSETID = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
+	// TODO: we currently support NSTETH only which has capped effective balance for one validator
+	// TODO: this is a bad practice, and for Lz, they have different version of endpoint with different chainID
+	// Do the validation before invoke oracle ralted functions instead of check these hard code ids here.
+	NSTETHAssetIDMainnet  = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x7595"
+	NSTETHAssetIDLocalnet = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
+	NSTETHAssetIDHolesky  = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x9d19"
+	NSTETHAssetIDSepolia  = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x9ce1"
```
Comment (MaxMustermann2) [inline x/oracle/keeper/native_token.go]: Can we convert this to an enum? I know Go doesn't support string enums but maybe via `type NSTAssetID string` and `const ETHMainnet NSTAssetID = "0x...."` ?

**Review by MaxMustermann2 — APPROVED** [review]

File: x/oracle/keeper/native_token.go (hunk 7 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go [hunk 8]
```diff
@@ -21,7 +21,27 @@ import (
...
+)
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go [hunk 9]
```diff
@@ -21,7 +21,26 @@ import (
 // msg(refund or slash on beaconChain): update staker's price, operator's price
 
 const (
-	NSTETHASSETID = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
+	// TODO: we currently support NSTETH only which has capped effective balance for one validator
+	// TODO: this is a bad practice, and for Lz, they have different version of endpoint with different chainID
+	// Do the validation before invoke oracle ralted functions instead of check these hard code ids here.
+	NSTETHAssetIDMainnet = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
+	NSTETHAssetIDHolesky = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x9d19"
+	NSTETHAssetIDSepolia = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x9ce1"
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: temporary solution

File: x/oracle/keeper/native_token.go [hunk 10]
```diff
@@ -20,8 +20,32 @@ import (
 // undelegate: update operator's price, operator's totalAmount, operator's totalShare, staker's share
 // msg(refund or slash on beaconChain): update staker's price, operator's price
 
+type NSTETHAssetID string
+
 const (
-	NSTETHASSETID = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_0x65"
+	// TODO: we currently support NSTETH only which has capped effective balance for one validator
+	// TODO: this is a bad practice, and for Lz, they have different version of endpoint with different chainID
+	// Do the validation before invoke oracle ralted functions instead of check these hard code ids here.
```
Comment (TimmyExogenous) [inline x/oracle/keeper/native_token.go]: typo: ralted

**Review by TimmyExogenous — APPROVED** [review]

**Review by mikebraver — APPROVED** [review]

File: x/oracle/keeper/native_token.go [hunk 11]
```diff
@@ -392,3 +414,12 @@ func parseBalanceChange(rawData []byte, sl types.StakerList) (map[string]int, er
...
 }
+
+// IsLimitChangesNST returns that is input assetID corresponding to asset which balance change has a cap limit
+func IsLimitedChangeNST(assetID string) bool {
+	return limitedChangeNST[NSTETHAssetID(assetID)]
+}
+
+func maxEffectiveBalance(assetID string) int {
+	return maxEffectiveBalances[NSTETHAssetID(assetID)]
+}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go:425]: same results

**Pull request merged**